### PR TITLE
[WIP] Add support for sparse arrays in fit, predict and evaluate.

### PR DESCRIPTION
**Pull request by gabrieldemarmiesse**
Hi!
This is the pull request for the "New additions" in the Requests for Contributions.

## The idea:
- When giving sparse arrays to keras in fit, predict or evaluate, if the corresponding Input layer wasn't specified as sparse when creating the model, the backend would complain about it.

## Request for comments

- My pull request only works for input data. If the target of the model is a sparse array, the backend will still complain . This is because of how I check if a conversion is necessary. See "modifications" for explanations. Ideas to solve this are welcome.

- If steps_per_epoch (or steps in predict and evaluate) is set, I noticed that the variable "ins" isn't sliced and never changes but the model is trained with it multiple times (it felt strange). So I applied the conversion before the loop to make it faster. I don't really understand if I'm wrong or not about "ins" never changing. Please be careful when reviewing this part.

## Modifications
- Before any loop, I create a list containing the indices of the arrays which need to a conversion to dense.
I loop though this list before calling f. 
- So in the case that the list is empty (no conversion needed) is still fast because the loop doesn't run a single iteration. It could be faster if we were just checking a boolean before the for loop, but it would bloat the code a bit.
- I use the attribute "sparse" of the corresponding input layer to know if the array needs to be converted or not.
- I made 3 test functions (one for fit, one for predict and one for evaluate). I didn't change the previous test function checking sparse arrays with Tensorflow.

Thank you.

### CODE REVIEW

File: tests/keras/engine/test_training.py [hunk 1]
```diff
@@ -497,6 +497,38 @@ def test_sparse_input_validation_split():
     model.fit(test_input, test_output, epochs=1, batch_size=2, validation_split=0.2)
 
 
+@keras_test
+def test_sparse_input_fit():
+    test_input = sparse.random(6, 3, density=0.25).tocsr()
+    in1 = Input(shape=(3,))
+    out1 = Dense(4)(in1)
+    test_output = np.random.random((6, 4))
+    model = Model(in1, out1)
```
Comment (fchollet) [inline tests/keras/engine/test_training.py]: Please test a multi-input, multi-output model.

File: keras/engine/training.py [hunk 2]
```diff
@@ -1157,6 +1158,17 @@ def _fit_loop(self, f, ins, out_labels=None, batch_size=None,
         for cbk in callbacks:
             cbk.validation_data = val_ins
 
+        # To prevent a slowdown, we find beforehand the arrays that need conversion.
+        indices_for_conversion_to_dense = []
+        for i in range(len(self._feed_input_names)):
```
Comment (fchollet) [inline keras/engine/training.py]: There is a problem here. `_feed_input_names` is the list of model *inputs*, but `ins` refers to list of the Keras function input placeholders. Typically, ins = model_inputs + model_outputs + sample_weights. Your setup will still work, but the conversion doesn't get applied to targets.

**Comment by gabrieldemarmiesse** [comment]
Thank you very much for the feedback! I will work on it.

**Comment by gabrieldemarmiesse** [comment]
That was fast! Thanks!

**Pull request merged**